src/lib/libast/man/stk.3
.fp 5 CW
.TH STK 3
.SH NAME
\fBstk\fR \- data stack storage library
.SH SYNOPSIS
.ta .75i 1.5i 2.25i 3i 3.75i 4.5i 5.25i 6i
.de Ss
.fl
.ne 3
.SS "\\$1"
..
.PP
.nf
\f3
#include <stk.h>

malloc_attr Stk_t *stkopen(int \fIflags\fP);
void stkoverflow(Stk_t *\fIstack\fP, void *(\fIoverflow\fP)(size_t));
Stk_t *stkinstall(Stk_t *\fIstack\fP, char *(\fIoverflow\fP)(size_t));
int stkclose(Stk_t *\fIstack\fP);
unsigned int stklink(Stk_t *\fIstack\fP)

void *stkalloc(Stk_t *\fIstack\fP, size_t \fIsize\fP);
char *stkcopy(Stk_t *\fIstack\fP, const char *\fIstring\fP);
void *stkset(Stk_t *\fIstack\fP, void *\fIaddress\fP, ptrdiff_t \fIoffset\fP);

void *stkseek(Stk_t *\fIstack\fP, ptrdiff_t \fIoffset\fP);
ptrdiff_t stktell(Stk_t *\fIstack\fP);
char *stkptr(Stk_t *\fIstack\fP, ptrdiff_t \fIoffset\fP);
void *stkfreeze(Stk_t *\fIstack\fP, size_t \fIextra\fP);
\fR
.fi
.SH DESCRIPTION
.PP
\f3stk\fP is a package of routines designed to provide efficient
stack oriented dynamic storage.
A stack abstraction consists of an ordered list of contiguous
memory regions, called stack frames, that can hold objects of
arbitrary size.
A stack is represented by the type \f3Stk_t\fP
defined in the header \f3<stk.h>\fP.
The type \f3Stk_t\fP is compatible with the type \f3Sfio_t\fP
defined by the \f3sfio\fP(3) library.
There is a preset initial stack which can be referenced
by the constant \f3stkstd\fP (stack standard).
.PP
Objects of variable size can be
added to a stack,
and programs can reference these objects directly with pointers.
In addition, the last object on the stack
(referred to here as the current object)
can be built incrementally.
The current object has an associated offset that determines its
current size.
While the current object is being built incrementally,
its location might
change so that it is necessary to reference the object with
relative offsets ranging from zero to the current offset of the object.
.PP
The \f3sfio\fP(3) output functions can be used to build
the current object incrementally.
An object that is built incrementally on the stack will
always occupy contiguous memory within a stack frame,
but until \f3stkfreeze\fP() is called,
the location in memory for the object can change.
There is a current offset associated with the current object that
determines where subsequent operations apply.
Initially, this offset is zero, and the offset changes as a result
of the operations you specify.
.Ss "Descriptions of the functions"
.PP
The \f3stkopen\fP() function creates a new stack.
The \fIflags\fP argument is an options bitmask.
If the \f3STK_SMALL\fP bit is set, the stack allocates memory in
small blocks, optimizing memory usage at the expense of performance.
If the \f3STK_NULL\fP bit is set, a stack overflow will cause stack
operations to return a null pointer instead of throwing an exception.
If the \f3STK_NULL\fP bit is not set and the \f3stkstd\fP stack has
a custom overflow function set (see \f3stkoverflow\fP() below),
then the new stack inherits the pointer to that overflow function.
If successful,
\f3stkopen\fP() returns a pointer to a stack whose reference
count is 1.
Otherwise, \f3stkopen\fP() returns a null pointer.
.PP
The \f3stkoverflow\fP() function specifies a function that will
be called whenever \f3malloc\fP(3) fails while trying to grow the
stack.
The \fIoverflow\fP function will be called with the size that was passed
to \f3malloc\fP(3).
The \fIoverflow\fP function can call \f3exit\fP(3), call \f3longjmp\fP(3)
or return.
If the \f2overflow\fP function returns,
it must return a pointer to a memory region of the given size.
If \f2overflow\fP is \f3NULL\fP, then the stack overflow action
is reset to the default.
If the stack was opened with the \f3STK_NULL\fP option,
the default is to return a null pointer,
otherwise the default is to write an error to standard error and to
call \f3exit\fP(2) with a non-zero exit value.
.PP
The \f3stkinstall\fP() function
is deprecated and provided for backward compatibility.
\f3stkinstall\fP() makes \f3stkstd\fP refer to the specified \fIstack\fP pointer
and returns a pointer to the previous \f3stkstd\fP stack.
If the \fIoverflow\fP argument is not null and the stack was not opened with
the \f3STK_NULL\fP option, then \fIoverflow\fP sets the overflow action
(see \f3stkoverflow\fP() above).
When \fIstack\fP is a null pointer,
the \f3stkstd\fP stack is not changed
but the \fIoverflow\fP function for the \f3stkstd\fP stack can be changed
and a pointer to the \f3stkstd\fP stack is returned.
(To replace deprecated \f3stkinstall\fP() usage, simply
pass the pointer to the desired stack to the \f3stk\fP functions,
and use \f3stkoverflow\fP() to set the stack overflow function).
.PP
The \f3stkclose\fP() function decrements the reference count and
frees the memory associated with
the specified stack
when the reference count is zero.
The effect of subsequent references to objects
on the stack is undefined.
.PP
The \f3stklink\fP() function increases the reference count for the
given \fIstack\fP and returns the increased count.
.PP
The
\f3stkalloc\fP() function returns an aligned pointer to space on the
active stack that can be used to hold any object of the given \fIsize\fP.
\f3stkalloc\fP() is similar to \f3malloc\fP(3) except that individual
items returned by \f3stkalloc\fP() can not be freed.
\f3stkalloc\fP() causes the offset of the current object to be set to
zero.
.PP
The
\f3stkcopy\fP() function copies the given string onto the stack
and returns a pointer to the \fIstring\fP on the stack.
\f3stkcopy\fP() causes the offset of the current object to be set to
zero.
.PP
The \f3stkset\fP() function finds the frame containing the given
\fIaddress\fP, frees all frames that were created after the one containing
the given \fIaddress\fP, and sets the current object to the given
\fIaddress\fP.
The top of the current object is set to \fIoffset\fP bytes from
current object.
If \fIaddress\fP is null, the stack is reset to the beginning.
If it is non-null, but is not the address of an object on the
stack, the program aborts and dumps core.
.PP
The \f3stkseek\fP() function is used set the offset for the
current object.
The \fIoffset\fP argument to \f3stkseek\fP() specifies the new
offset for the current object.
The frame will be extended or moved
if \f3offset\fP causes the new current offset to extend beyond the
current frame.
\f3stkseek\fP() returns a pointer to the beginning of the current object.
.PP
The \f3stktell\fP() function gives the offset of the current object.
.PP
The \f3stkptr\fP() function converts the given \f3offset\fP
for the current object into a memory address on the stack.
This address is only valid until another stack operation is given.
The result is not defined if \fIoffset\fP exceeds the size of the current
object.
This function returns a \f3char\fP pointer instead of \f3void\fP pointer
to make it easy to dereference the return value directly and read
an individual byte on the stack.
.PP
The \f3stkfreeze\fP()
function terminates the current object on the
stack and returns a pointer to the beginning of this object.
If \fIextra\fP is non-zero, \fIextra\fP bytes are added to the stack
before the current object is terminated.  The first added byte will
contain zero and the contents of the remaining bytes are undefined.
.SH HISTORY
The
\f3stk\fP
interface was derived from similar routines in the KornShell code
that is used for building parse trees and carrying out expansions.
It provides an efficient mechanism for grouping dynamically allocated
objects so that they can be freed all at once rather than individually.
.PP
In 2024, as part of changes made by
Martijn Dekker <martijn@inlv.org> for ksh 93u+m,
the return types of \f3stkalloc\fP(), \f3stkset\fP(), \f3stkseek\fP() and
\f3stkfreeze\fP() were changed from \f3char*\fP to \f3void*\fP,
the \f3stkoverflow\fP() function was added,
and the \f3stkinstall\fP() function was deprecated.
.SH AUTHOR
David Korn
.SH SEE ALSO
\f3exit\fP(2),
\f3longjmp\fP(3),
\f3malloc\fP(3),
\f3sfio\fP(3)


NEWS
This documents significant changes in the dev branch of ksh 93u+m.
For full details, see the git log at: https://github.com/ksh93/ksh
Uppercase BUG_* IDs are shell bug IDs as used by the Modernish shell library.

2026-02-04:

- Fixed a severe arithmetic bug (which thankfully never made it into a release
  version) introduced on 2024-11-30, in which 'typeset -i' variables were
  incorrectly treated as unsigned int when rendering their values as strings
  (e.g. 'typeset -i foo=-1; echo $foo' output 4294967295 instead of -1).

- Fixed a bug that caused ksh to set the underlying locale to
  C rather than C.UTF-8 on systems with a native C.UTF-8 locale.

- Improved the handling of multibyte characters in the ksh implementation
  of the C.UTF-8 locale on older operating systems that don't provide a
  native version.

2026-02-01:

- Replaced the 2021-03-09 fix for ${!foo@} and ${!foo*}, which was not quite
  correct, with a new fix that also corrects ${#foo@} and ${#foo*}; these now
  also count 'foo' itself as a variable whose name begins with 'foo'.

- Fixed a possible out of bounds crash that could occur when running
  a POSIX function (aka a function of the form 'name() {').

2026-01-28:

- [v1.1] Backported 'builtin -p' and 'umask -p' from ksh93v-, which
  can be used to print reinputtable commands for recreating the builtin
  table state and umask.

- [v1.1] Backported 'trap -l' from ksh93v- for listing signals.

- Fixed a regression, introduced with the printf fixes on 2023-05-18, that
  broke date specs relative to now, such as: printf '%T\n' 'in 10 weeks'

- Fixed incorrect behaviour that occurred when IFS, PATH, SHELL,
  FPATH, CDPATH, SECONDS, ENV, LANG, LC_ALL, LC_CTYPE, LC_MESSAGES,
  LC_COLLATE, LC_NUMERIC, or LC_TIME were declared as a local
  variable within a function using 'typeset' while a custom shell
  discipline function was defined for them in the global scope.

- Fixed memory fault when attempting to run /dev/fd/1 as a script.

- Fixed the value of $0 when running a /dev/fd file descriptor as a script.

2026-01-21:

- Fixed several IFS field splitting bugs in read(1) that occurred
  when it is given a null delimiter (read -d '').

2025-09-30:

- The rm path-bound built-in command from libcmd no longer suppresses error
  messages when the -f/--force option is used.

2025-06-14:

- Fixed a bug occurring on systems with posix_spawn(3), spawnve(2), and
  spawn(2), which when exploited could cause ksh93 to execute scripts
  without a shebang with a shell other than itself, provided /bin/ksh
  was hijacked, or in the case of ksh being unable to obtain it's own
  executable path, whatever the SHELL variable was set to.

- Fixed a bug that could cause scripts without a shebang to run
  without POSIX mode enabled when executed in a virtual subshell.

- Fixed a bug that could allow an attacker to hijack ksh93 during
  shell initialization to execv a shell other than ksh itself, leading to
  the ksh process being surreptitiously replaced with bash, zsh, or
  really anything, provided the binary name ends with the letters
  'sh'.

2025-05-31:

- Fixed a bug that caused the multiline option to break when PS*
  prompts use the kitty terminal's escape sequence (e.g. $'\E[4:4m'
  in the PS1 prompt for formatting spaced dots).

2025-05-29:

- Fixed a serious but rarely occurring regression that corrupts variable
  names in assignments. It could be triggered in large scripts and cause a
  spurious "invalid variable name" error. Bug introduced on 2024-01-16.

- Restored the ability to import environment variables with multibyte UTF-8
  names while in a UTF-8 locale. This broke on 2020-09-26.

2025-05-27:

- Fixed a crash that could occur in some scenarios when /opt/ast/bin
  builtins handled invalid options in virtual subshells.

- Fixed a bug in printf where \u0025 was taken as a % operator in a
  formatter argument. For example, printf '\u0025\n' now correctly prints a
  % followed by a newline instead of producing an error.

- [v1.1] Fixed another bug with superfluous operands in test/[ that slipped
  through the cracks on 2024-12-08: test -n -a "" incorrectly returned 0. In
  native mode, this is now an error (superfluous operand to the unary -n
  operator) and in POSIX mode, it returns 1/false, as POSIX-1.2017 rules
  require this to be a test for the non-emptiness of both strings -n and "".

2025-05-26:

- Fixed a POSIX mode bug in test/[ (with 'posix' shell option on):
  * test ! foo -o bar	wrongly returned 0/true, now returns 1/false.
  * test ! foo -a ''	wrongly returned 1/false, now returns 0/true.
  This regression was introduced on 2021-11-13.

- Made the handling of deprecated parenthesis operators in test/[ slightly
  more permissive for increased compatibility with other shells and external
  test/[ commands.

2025-05-15:

- Fixed an issue that prevented Unicode code points in $'...' strings and
  printf(1) formatters from being converted to the current locale. For
  example, if the locale on Linux/BSD/macOS is el_GR.ISO8859-7, echo $'\u03B1'
  now outputs a Greek lower-case alpha encoded in ISO8859-7 ($'\xE1').

2025-05-10:

- Fixed a memory leak in history expansion.

- Fixed a crash in 'fc -p'/'hist -p' that occurred when an incorrect history
  expansion followed a correct one in the command operands.

2025-05-08:

- [v1.1] Fixed: case modification expansions (added on 2025-04-23) were not
  properly subject to field splitting and pathname expansion if unquoted.

2025-05-05:

- [v1.1] grep(1) has been added as a path-bound built-in command. It is now
  available by default on ksh linked against dynamic libcmd, and may
  optionally be included in ksh linked against static libcmd. This grep
  supports eight kinds of regex, including sh and ksh glob patterns.

2025-05-02:

- [v1.1] Self-documentation output from any command's --man, --help, and
  other options is no longer written to standard error with exit status 2,

src/lib/libast/misc/stk.c
/***********************************************************************
*                                                                      *
*               This software is part of the ast package               *
*          Copyright (c) 1985-2012 AT&T Intellectual Property          *
*          Copyright (c) 2020-2026 Contributors to ksh 93u+m           *
*                      and is licensed under the                       *
*                 Eclipse Public License, Version 2.0                  *
*                                                                      *
*                A copy of the License is available at                 *
*      https://www.eclipse.org/org/documents/epl-2.0/EPL-2.0.html      *
*         (with md5 checksum 84283fa8859daf213bdda5a9f8d1be1d)         *
*                                                                      *
*                 Glenn Fowler <gsf@research.att.com>                  *
*                  David Korn <dgk@research.att.com>                   *
*                   Phong Vo <kpv@research.att.com>                    *
*                  Martijn Dekker <martijn@inlv.org>                   *
*            Johnothan King <johnothanking@protonmail.com>             *
*                                                                      *
***********************************************************************/
/*
 *   Routines to implement a stack-like storage library
 *
 *   A stack consists of a link list of variable size frames
 *   The beginning of each frame is initialized with a frame structure
 *   that contains a pointer to the previous frame and a pointer to the
 *   end of the current frame.
 *
 *   This is a rewrite of the stk library that uses sfio
 *
 *   David Korn
 *   AT&T Research
 *   dgk@research.att.com
 *
 */

#include	<sfio_t.h>
#include	<ast.h>
#include	<align.h>
#include	<stk.h>

/*
 *  A stack is a header and a linked list of frames
 *  The first frame has structure
 *	Sfio_t
 *	Sfdisc_t
 *	struct stk
 * Frames have structure
 *	struct frame
 *	data
 */

#define STK_ALIGN	((size_t)ALIGN_BOUND)
#define STK_FSIZE	(1024*sizeof(char*))
#define STK_HDRSIZE	(sizeof(Sfio_t)+sizeof(Sfdisc_t))

typedef void* (*_stk_overflow_)(size_t);
typedef char* (*_old_stk_overflow_)(size_t);	/* for stkinstall (deprecated) */

static int stkexcept(Sfio_t*,int,void*,Sfdisc_t*);
static Sfdisc_t stkdisc = { 0, 0, 0, stkexcept };

Sfio_t	_Stak_data = SFNEW(NULL,0,-1,SFIO_STATIC|SFIO_WRITE|SFIO_STRING,&stkdisc);

struct frame
{
	char	*prev;		/* address of previous frame */
	char	*end;		/* address of end this frame */
	char	**aliases;	/* address aliases */
	ssize_t	nalias;	/* number of aliases */
};

struct stk
{
	_stk_overflow_	stkoverflow;	/* called when malloc fails */
	unsigned int	stkref;		/* reference count */
	int		stkflags;	/* stack attributes */
	char		*stkbase;	/* beginning of current stack frame */
	char		*stkend;	/* end of current stack frame */
};

static size_t		init;		/* 1 when initialized */
static struct stk	*stkcur;	/* pointer to current stk */
static char		*stkgrow(Sfio_t*, size_t);

#define stream2stk(stream)	((stream)==stkstd? stkcur:\
				 ((struct stk*)(((char*)(stream))+STK_HDRSIZE)))
#define stk2stream(sp)		((Sfio_t*)(((char*)(sp))-STK_HDRSIZE))
#define stkleft(stream)		((stream)->_endb-(stream)->_data)

static const char Omsg[] = "out of memory while growing stack\n";

/*
 * default overflow exception
 */
static cold noreturn void *overflow(size_t n)
{
	NoP(n);
	write(2,Omsg, sizeof(Omsg)-1);
	exit(128);
	UNREACHABLE();
}

/*
 * initialize stkstd, sfio operations may have already occurred
 */
static cold void stkinit(size_t size)
{
	Sfio_t *sp;
	init = size;
	sp = stkopen(0);
	init = 1;
	stkinstall(sp,(_old_stk_overflow_)overflow);
}

static int stkexcept(Sfio_t *stream, int type, void* val, Sfdisc_t* dp)
{
	NoP(dp);
	NoP(val);
	switch(type)
	{
	    case SFIO_CLOSING:
		{
			struct stk *sp = stream2stk(stream);
			char *cp = sp->stkbase;
			struct frame *fp;
			if(--sp->stkref == 0)
			{
				if(stream==stkstd)
					stkset(stream,NULL,0);
				else
				{
					while(1)
					{
						fp = (struct frame*)cp;
						if(fp->prev)
						{
							cp = fp->prev;
							free(fp);
						}
						else
						{
							free(fp);
							break;
						}
					}
				}
			}
			stream->_data = stream->_next = 0;
		}
		return 0;
	    case SFIO_FINAL:
		free(stream);
		return 1;
	    case SFIO_DPOP:
		return -1;
	    case SFIO_WRITE:
	    case SFIO_SEEK:
		{
			size_t size = (size_t)sfvalue(stream);
			if(init)
			{
				Sfio_t *old = 0;
				if(stream!=stkstd)
					old = stkinstall(stream,NULL);
				if(!stkgrow(stkstd,size-(size_t)(stkstd->_endb-stkstd->_data)))
					return -1;
				if(old)
					stkinstall(old,NULL);
			}
			else
				stkinit(size);
		}
		return 1;
	    case SFIO_NEW:
		return -1;
	}
	return 0;
}

/*
 * create a stack
 */
malloc_attr Sfio_t *stkopen(int flags)
{
	size_t bsize;
	Sfio_t *stream;
	struct stk *sp;
	struct frame *fp;
	Sfdisc_t *dp;
	char *cp;
	if(unlikely(!(stream=newof(NULL,Sfio_t, 1, sizeof(*dp)+sizeof(*sp)))))
		return NULL;
	dp = (Sfdisc_t*)(stream+1);
	dp->exceptf = stkexcept;
	sp = (struct stk*)(dp+1);
	sp->stkref = 1;
	sp->stkflags = flags;
	if(flags&STK_NULL) sp->stkoverflow = 0;
	else sp->stkoverflow = stkcur?stkcur->stkoverflow:overflow;
	bsize = init+sizeof(struct frame);
	if(flags&STK_SMALL)
		bsize = roundof(bsize,STK_FSIZE/16);
	else
		bsize = roundof(bsize,STK_FSIZE);
	bsize -= sizeof(struct frame);
	if(unlikely(!(fp=newof(NULL,struct frame, 1,bsize))))
	{
		free_sized(stream,sizeof(Sfio_t)*1+sizeof(*dp)+sizeof(*sp));
		return NULL;
	}
	cp = (char*)(fp+1);
	sp->stkbase = (char*)fp;
	fp->prev = 0;
	fp->nalias = 0;
	fp->aliases = 0;
	fp->end = sp->stkend = cp+bsize;
	if(unlikely(!sfnew(stream,cp,bsize,-1,SFIO_STRING|SFIO_WRITE|SFIO_STATIC|SFIO_EOF)))
		return NULL;
	sfdisc(stream,dp);
	return stream;
}

/*
 * return a pointer to the current stack
 * if <stream> is not null, it becomes the new current stack
 * <oflow> becomes the new overflow function
 */
Sfio_t *stkinstall(Sfio_t *stream, _old_stk_overflow_ oflow)
{
	Sfio_t *old;
	struct stk *sp;
	if(!init)
	{
		stkinit(1);
		if(oflow)
			stkcur->stkoverflow = (_stk_overflow_)oflow;
		return NULL;
	}
	old = stkcur?stk2stream(stkcur):0;
	if(likely(stream))
	{
		sp = stream2stk(stream);
		while(sfstack(stkstd, SFIO_POPSTACK));
		if(likely(stream!=stkstd))
			sfstack(stkstd,stream);
		stkcur = sp;
	}
	else
		sp = stkcur;
	if(likely(oflow))
		sp->stkoverflow = (_stk_overflow_)oflow;
	return old;
}

/*
 * set or unset the overflow function
 */
void stkoverflow(Sfio_t *stream, _stk_overflow_ oflow)
{
	struct stk *sp;
	if(!init)
		stkinit(1);
	sp = stream2stk(stream);
	sp->stkoverflow = oflow ? oflow : (sp->stkflags & STK_NULL ? NULL : overflow);
}

/*
 * increase the reference count on the given <stack>
 */
unsigned int stklink(Sfio_t* stream)
{
	struct stk *sp = stream2stk(stream);
	return sp->stkref++;
}

/*
 * terminate a stack and free up the space
 * >0 returned if reference decremented but still > 0
 *  0 returned on last close
 * <0 returned on error
 */
int stkclose(Sfio_t* stream)
{
	struct stk *sp = stream2stk(stream);
	if(sp->stkref>1)
	{
		sp->stkref--;
		return 1;
	}
	return sfclose(stream);
}

/*
 * reset the bottom of the current stack back to <address>
 * if <address> is null, then the stack is reset to the beginning
 * if <address> is not in this stack, the program dumps core
 * otherwise, the top of the stack is set to stkbot+<offset>
 */
void *stkset(Sfio_t *stream, void *address, ptrdiff_t offset)
{
	struct stk *sp = stream2stk(stream);
	char *cp, *loc = (char*)address;
	struct frame *fp;
	unsigned int frames = 0;
	ssize_t n;
	if(!init)
		stkinit((size_t)(offset+1));
	while(1)
	{
		fp = (struct frame*)sp->stkbase;
		cp = sp->stkbase + roundof(sizeof(struct frame), STK_ALIGN);
		n = fp->nalias;
		while(unlikely(n-->0))
		{
			if(loc==fp->aliases[n])
			{
				loc = cp;
				break;
			}
		}
		/* see whether <loc> is in current stack frame */
		if(likely(loc>=cp && loc<=sp->stkend))
		{
			if(unlikely(frames))  /* acc. gcov */
				sfsetbuf(stream,cp,(size_t)(sp->stkend-cp));
			stream->_data = (unsigned char*)(cp + roundof((size_t)(loc-cp),STK_ALIGN));
			stream->_next = (unsigned char*)loc+offset;
			goto found;
		}
		if(fp->prev)
		{
			sp->stkbase = fp->prev;
			sp->stkend = ((struct frame*)(fp->prev))->end;
			free(fp);
		}
		else
			break;
		frames++;
	}
	/* not found: produce a useful stack trace now instead of a useless one later */
	if(unlikely(loc))
		abort();
	/* set stack back to the beginning */
	cp = (char*)(fp+1);
	if(unlikely(frames))  /* acc. gcov */
		sfsetbuf(stream,cp,(size_t)(sp->stkend-cp));
	else
		stream->_data = stream->_next = (unsigned char*)cp;
found:
	return stream->_data;
}

/*
 * allocate <n> bytes on the current stack
 */
hot void *stkalloc(Sfio_t *stream, size_t n)
{
	unsigned char *old;
	if(!init)
		stkinit(n);
	n = roundof(n,STK_ALIGN);
	if(unlikely(stkleft(stream) <= (ssize_t)n) && unlikely(!stkgrow(stream,n)))
		return NULL;
	old = stream->_data;
	stream->_data = stream->_next = old+n;
	return old;
}

/*
 * begin a new stack word of at least <n> bytes
 */
hot void *_stkseek(Sfio_t *stream, ptrdiff_t n)
{
	if(!init)
		stkinit((size_t)n);
	if(unlikely(stkleft(stream) <= n) && unlikely(!stkgrow(stream,(size_t)n)))
		return NULL;
	stream->_next = stream->_data+n;
	return stream->_data;
}

/*
 * advance the stack to the current top
 * if extra is non-zero, first add extra bytes and zero the first
 */
hot void *stkfreeze(Sfio_t *stream, size_t extra)
{
	unsigned char *old, *top;
	if(!init)
		stkinit(extra);
	old = stream->_data;
	top = stream->_next;
	if(extra)
	{
		if(unlikely((ssize_t)extra > (stream->_endb-stream->_next)))
		{
			if (!(top = (unsigned char*)stkgrow(stream,extra)))
				return NULL;
			old = stream->_data;
		}
		*top = 0;
		top += extra;
	}
	stream->_next = stream->_data += roundof((size_t)(top-old),STK_ALIGN);
	return (char*)old;
}

/*
 * copy string <str> onto the stack as a new stack word
 */
char	*stkcopy(Sfio_t *stream, const char* str)
{
	unsigned char *cp = (unsigned char*)str;
	size_t n;
	size_t off = (size_t)stktell(stream);
	char buff[40], *tp=buff;
	if(unlikely(off))
	{
		if(off > sizeof(buff))
		{
			if(unlikely(!(tp = malloc(off))))
			{
				struct stk *sp = stream2stk(stream);
				if(!sp->stkoverflow || !(tp = (*sp->stkoverflow)(off)))
					return NULL;
			}
		}
		memcpy(tp, stream->_data, off);
	}
	while(*cp++);
	n = roundof((size_t)(cp-(unsigned char*)str),STK_ALIGN);
	if(!init)
		stkinit(n);
	if(unlikely(stkleft(stream) <= (ssize_t)n) && unlikely(!stkgrow(stream,n)))
		cp = 0;
	else
	{
		strcpy((char*)(cp=stream->_data),str);
		stream->_data = stream->_next = cp+n;
		if(unlikely(off))
		{
			_stkseek(stream,(ptrdiff_t)off);
			memcpy(stream->_data, tp, off);
		}
	}
	if(unlikely(tp!=buff))
		free(tp);
	return (char*)cp;
}

/*
 * add a new stack frame of size >= <n> to the current stack.
 * if <n> > 0, copy the bytes from stkbot to stktop to the new stack
 * if <n> is zero, then copy the remainder of the stack frame from stkbot
 * to the end is copied into the new stack frame
 */

static char *stkgrow(Sfio_t *stream, size_t size)
{
	size_t n = size;
	struct stk *sp = stream2stk(stream);
	struct frame *fp= (struct frame*)sp->stkbase;
	char *cp, *dp=0;
	size_t m = (size_t)stktell(stream);
	ptrdiff_t endoff;
	char *end=0, *oldbase=0;
	ssize_t nn=0,add=1;
	n += (m + sizeof(struct frame)+1);
	if(sp->stkflags&STK_SMALL)
		n = roundof(n,STK_FSIZE/16);
	else
		n = roundof(n,STK_FSIZE);
	/* see whether current frame can be extended */
	if(unlikely(stkptr(stream,0)==sp->stkbase+sizeof(struct frame)))
	{
		nn = fp->nalias+1;
		dp=sp->stkbase;
		sp->stkbase = ((struct frame*)dp)->prev;
		end = fp->end;
		oldbase = dp;
	}
	endoff = end - dp;
	cp = newof(dp, char, n, (size_t)nn*sizeof(char*));
	if(unlikely(!cp) && unlikely(!sp->stkoverflow || !(cp = (*sp->stkoverflow)(n))))
		return NULL;
	if(unlikely(dp==cp))
	{
		nn--;
		add = 0;
	}
	else if(unlikely(dp))
	{
		dp = cp;
		end = dp + endoff;
	}
	fp = (struct frame*)cp;
	fp->prev = sp->stkbase;
	sp->stkbase = cp;
	sp->stkend = fp->end = cp+n;
	cp = (char*)(fp+1);
	cp = sp->stkbase + roundof((size_t)(cp-sp->stkbase),STK_ALIGN);
	if(unlikely(fp->nalias=nn))
	{
		fp->aliases = (char**)fp->end;
		if(end && nn>add)
			memmove(fp->aliases,end,(size_t)(nn-add)*sizeof(char*));
		if(add)
			fp->aliases[nn-1] = oldbase + roundof(sizeof(struct frame),STK_ALIGN);
	}
	if(m && !dp)
		memcpy(cp,(char*)stream->_data,m);
	sfsetbuf(stream,cp,(size_t)(sp->stkend-cp));
	return (char*)(stream->_next = stream->_data+m);
}
